Give all model objects with an id a __repr__ that includes the id

natefoo opened the pull request:
Useful when developing, but it also helps when an error message includes the string representation of an object.

Before:
```
<galaxy.model.Job object at 0x7f7ee5b71f10>
```

After:
```
<galaxy.model.Job(None)>
<galaxy.model.Job(1)>
```

natefoo committed:
Give all model objects with an id a __repr__ that includes the id

nsoranzo commented on lib/galaxy/model/__init__.py:
```diff
@@ -4544,12 +4550,12 @@ def to_dict(self, view='collection', value_mapper=None):
 
 
-class WorkflowInvocationStepJobAssociation(Dictifiable):
+class WorkflowInvocationStepJobAssociation(Dictifiable, RepresentById):
```
I just found out this class can be removed, it was introduced in commit d5f98f49045bca351e982f17cac9386308089eeb , but its use was removed in commit e82012bde154c456f24743f3b61c0ed9ffc7eca2 , both part of PR #5013.

nsoranzo commented on lib/galaxy/model/__init__.py:
```diff
@@ -4544,12 +4550,12 @@ def to_dict(self, view='collection', value_mapper=None):
 
 
-class WorkflowRequest(Dictifiable):
+class WorkflowRequest(Dictifiable, RepresentById):
```
Also this class can be removed, introduced in commit d1cd4ab4d28c72f365852b75e2f2026c6b76ebbd and never used.

nsoranzo commented on lib/galaxy/model/__init__.py:
```diff
@@ -4587,45 +4593,45 @@ def __init__(self, workflow_step=None, name=None, value=None):
 
 
-class WorkflowRequestInputStepParmeter(Dictifiable):
+class WorkflowRequestInputStepParmeter(Dictifiable, RepresentById):
```
Uh, this is misspelled (s/Parmeter/Parameter/) both in this file and in lib/galaxy/model/mapping.py

nsoranzo commented on lib/galaxy/model/__init__.py:
```diff
@@ -5089,18 +5095,18 @@ def __init__(self, state=None, path=None, info=None, pid=None, socket=None, para
 
 
-class ItemTagAssociation(Dictifiable):
+class ItemTagAssociation(Dictifiable, RepresentById):
```
That's a base class as `BaseJobMetric` and `ItemRatingAssociation`, so I think this can be reverted.

nsoranzo commented on lib/galaxy/model/__init__.py:
```diff
@@ -5176,7 +5182,7 @@ def __init__(self, id=None, user=None, tool_id=None, tag_id=None, user_tname=Non
 
 
-class WorkRequestTagAssociation(ItemTagAssociation):
+class WorkRequestTagAssociation(ItemTagAssociation, RepresentById):
```
Another class that can be removed, also introduced in commit d1cd4ab and never used.

jmchilton commented on lib/galaxy/model/__init__.py:
```diff
@@ -78,6 +78,11 @@
 
+class RepresentById(object):
+    def __repr__(self):
+        return '<galaxy.model.%s(%s)>' % (self.__class__.__name__, self.id)
```
This will fire a DB query after a flush for every log statement that includes a model object. Consider using cached_id in this file instead (https://github.com/galaxyproject/galaxy/pull/6524/files#diff-5805bf58afdc31ee4a9c4e788301b11dR163).

jmchilton commented:
I'm -0 on this, it feels a bit too heavy and log messages should be more carefully crafted. Job has a ``log_str`` method you can use instead that produces a nicer, more useful log line IMO.

natefoo commented:
They should be, but in cases where they're not, it really helps as an admin to know what object is causing the problems you're experiencing.

natefoo committed:
Use model.cached_id() instead of <obj>.id as per the tip from
@jmchilton. Also, include the object's id (address) as with the default
repr since it indicates whether an object is a different copy from
another with the same id (database).

Also catch any exceptions because failing to repr a model object could
be pretty bad.

nsoranzo commented:
This PR needs to be rebased.

martenson commented:
@jmchilton I assume nothing changed for the worse on your -0 with the latest commit: https://github.com/galaxyproject/galaxy/pull/6757/commits/414ce0a0eb18ad224d31136ad886079843814b66 therefore I am merging this.
